### ISSUE TITLE
this.hub is undefined in estree mode

### ISSUE THREAD

**Issue by coderaiser** [babel/babel#11350]
## Bug Report

This related to #8617

- [x] I would like to work on a fix!

**Current Behavior**
`@babel/parser` has a different behavior when `estree` enabled, and when not.

**Input Code**
When `estree` disabled code with `Duplicated variable declaration` throws while parsing:

```js
const {parse} = require('@babel/parser');
const traverse = require('@babel/traverse').default;

const code = `function square(n) {
  const n = 1;
}`;

// crash with an error: Identifier 'n' has already been declared (1:16)
const ast = parse(code);
```

When `estree` mode enabled code is parsed, but can't be traversed because of `this.hub` is undefined:

```js
const {parse} = require('@babel/parser');
const traverse = require('@babel/traverse').default;

const code = `function square(n) {
  const n = 1;
}`;

const ast2 = parse(code, {
    plugins: [
        'estree',
    ]
});

traverse(ast, {
  enter(path) {
      console.log('n');
  }
});

/*
    throw this.hub.buildError(id, `Duplicate declaration "${name}"`, TypeError);
                     ^

TypeError: Cannot read property 'buildError' of undefined
    at Scope.checkBlockScopedCollisions (/home/coderaiser/node_modules/@babel/traverse/lib/scope/index.js:402:22)
*/
```

**Expected behavior/code**
Would be great if in `estree` mode `babel` throws while parsing and/or can throw while traversing
about `Duplicate declaration`. 

**Environment**

- Babel version(s): `v7.9.4`
- Node/npm version: Node 12(13)/npm 6
- OS: `Ubuntu`
- Monorepo: no
- How you are using Babel: `parse`, `traverse`

**Possible Solution**

I suggest to use `TypeError`, because this is what [hub actually does](
https://github.com/babel/babel/blob/f6c7bf36cec81baaba8c37e572985bb59ca334b1/packages/babel-traverse/src/hub.js), this way:

```js
throw TypeError(`Duplicate declaration "${name}"`);
```

If of course `buildError` doesn't used in development purpose to set breakpoint on error building, anyways `Chrome Developer Tools` debugger already suggest similar feature `pause on exceptions` (caught and uncaught).

**Comment by JLHwung** [comment]
You can wrap the parsed ast into a `File` class (defined in `@babel/core`) so `this.hub` is available, like we did in #10575.

I admit that this workaround is not ideal as ultimately `babel/traverse` should be allowed to run as a standalone package. Historically there has been attempt on fixing this issue (#5050) but later reverted in #5306.

@coderaiser I think we can surely revisit this issue and see if we can re-land #5050. PR is welcome.

**Comment by rob-myers** [comment]
@JLHwung do you mind providing an example?

This code fails when my visitor uses `.getSource()`:
```tsx
import * as babel from '@babel/core';
// ...
const parsed = babel.parseSync(code, { filename, plugins });
const traversed = babel.traverse(parsed!, visitor);
```
How do I wrap things in a `File`?

**Comment by JLHwung** [comment]
@rob-myers Please don't rely on the internal file structures on `@babel/core`, they _could_ change. Instead you can import `File` from the wildcard `babel`

```js
import * as babel from '@babel/core';
const File = babel.File;
```

**Comment by Magin-CC** [comment]
same problem for me

**Comment by IanVS** [comment]
~`File` no longer seems to be exported by `@babel/core`.~ It just isn't in the typescript types.

**Comment by simbahax** [comment]
Hi! 
I would be open to work on this issue, but I am not sure what the best way forward would be.
`traverse` could wrap the node to traverse by default in a file (`new File({ filename: 'asterix://tmp' }, { ast: userSuppliedAst }).ast`, but that's rather dirty.

I am guessing that most people run into this while encountering an error during [`checkBlockScopedCollisions`](https://github.com/babel/babel/blob/4a2be37634b5f524ac43fb0fcdaed385c68b1f05/packages/babel-traverse/src/scope/index.ts#L633).
Would it be feasible to, at least for this scenario, fallback to something like this?
```
if (duplicate) {
      if (!this.path.hub) throw new TypeError(`Duplicate declaration "${name}"`)
      [...]
    }
```

**Comment by JLHwung** [comment]
The `File` wrapper is provided by the `@babel/core`, and added to every input source, such that the error message could provide important error location info: e.g. the line and the column of the input triggering errors.

If you construct the AST yourself, it is likely that the error location info is not very helpful. If for any reasons you don't want to use `@babel/core`, you can supply a minimal `Hub` and a minimal `File`:

```js
import * as t from "@babel/types";
import traverse, { Hub, NodePath } from "@babel/traverse";

traverse(ast, {
  enter(path) {
      console.log(path.type);
  }
}, undefined, undefined, new NodePath(new Hub(), t.file(ast)));
```

I think what we can do in a new PR is to figure out when we should provide a minimal `File` in the `traverse` function, it should be done only once such that new hubs are not created in a recursive traversal. A good rule may be to create the hub when the provided `ast` is a program and the `parentPath` is not defined.

**Comment by simbahax** [comment]
Thanks for the quick reply!

PR is ready for discussion: https://github.com/babel/babel/pull/17672 :)
